agent/dns.go

```diff
@@ -764,26 +764,43 @@ func (d *DNSServer) dispatch(remoteAddr net.Addr, req, resp *dns.Msg, maxRecursi
 			MaxRecursionLevel: maxRecursionLevel,
 			EnterpriseMeta:    locality.EnterpriseMeta,
 		}
-		// Support RFC 2782 style syntax
-		if n == 2 && strings.HasPrefix(queryParts[1], "_") && strings.HasPrefix(queryParts[0], "_") {
 
-			// Grab the tag since we make nuke it if it's tcp
-			tag := queryParts[1][1:]
+		tag := ""
 
-			// Treat _name._tcp.service.consul as a default, no need to filter on that tag
-			if tag == "tcp" {
-				tag = ""
+		// If the first two query parts begin with _, this is an RFC 2782 style SRV lookup.
+		if strings.HasPrefix(queryParts[0], "_") && strings.HasPrefix(queryParts[1], "_") && n >= 2 {
```

Comment:
Note: I changed the order of these if-conditional checks. I'm not sure why we'd check whether the 1st query part starts with `_` before checking the 0th.

I also changed `n == 2` to `n >= 2`. The `n == 2` codepath is unchanged. But the `n >= 3` codepath is new.